tools/walletextension/ratelimiter/rate_limiter.go

```diff
@@ -0,0 +1,215 @@
+			}
+		}
+	}
+	return uint32(totalComputeTime / time.Millisecond)
+}
+
+type RateLimiter struct {
+	mu                    sync.Mutex
+	users                 map[common.Address]*RateLimitUser
+	userComputeTime       uint32
+	window                uint32
+	maxConcurrentRequests uint32
+	totalRequests         uint64
+	rateLimitedRequests   uint64
+}
+
+// IncrementTotalRequests increments the total requests counter by 1 with thread safety.
+func (rl *RateLimiter) IncrementTotalRequests() {
+	rl.mu.Lock()
+	defer rl.mu.Unlock()
+	rl.totalRequests++
+}
+
+// IncrementRateLimitedRequests increments the total requests counter by 1 with thread safety.
+func (rl *RateLimiter) IncrementRateLimitedRequests() {
+	rl.mu.Lock()
+	defer rl.mu.Unlock()
+	rl.rateLimitedRequests++
+}
+
+// GetMaxConcurrentRequest returns the maximum number of concurrent requests allowed.
+func (rl *RateLimiter) GetMaxConcurrentRequest() uint32 {
+	rl.mu.Lock()
+	defer rl.mu.Unlock()
+	return rl.maxConcurrentRequests
+}
+
+// GetUserComputeTime returns the user compute time in milliseconds.
+func (rl *RateLimiter) GetUserComputeTime() uint32 {
+	rl.mu.Lock()
+	defer rl.mu.Unlock()
+	return rl.userComputeTime
+}
+
+func NewRateLimiter(rateLimitUserComputeTime uint32, rateLimitWindow uint32, concurrentRequestsLimit uint32) *RateLimiter {
+	rl := &RateLimiter{
+		users:                 make(map[common.Address]*RateLimitUser),
+		userComputeTime:       rateLimitUserComputeTime,
+		window:                rateLimitWindow,
+		maxConcurrentRequests: concurrentRequestsLimit,
+	}
+	go rl.logRateLimitedStats()
+	go rl.periodicPrune()
+	return rl
+}
+
+// Allow checks if the user is allowed to make a request based on the rate limit threshold
+// before comparing to the threshold also decays the score of the user based on the decay rate
+func (rl *RateLimiter) Allow(userID common.Address) (bool, uuid.UUID) {
+	// If the userComputeTime is 0, allow all requests (rate limiting is disabled)
+	if rl.GetUserComputeTime() == 0 {
+		return true, zeroUUID
+	}
+	// Increment the total requests counter for statistics
+	rl.IncrementTotalRequests()
+
+	// Check if the user has reached the maximum number of concurrent requests
+	if uint32(rl.CountOpenRequests(userID)) >= rl.GetMaxConcurrentRequest() {
+		rl.IncrementRateLimitedRequests()
+		return false, zeroUUID
+	}
+
+	// Check if user is in limits of rate limiting
+	userComputeTimeForUser := rl.SumComputeTime(userID)
+	if userComputeTimeForUser <= rl.userComputeTime {
+		requestUUID := rl.AddRequest(userID, RequestInterval{Start: time.Now()})
+		return true, requestUUID
+	}
+	rl.IncrementRateLimitedRequests()
+	return false, zeroUUID
+}
+
+// PruneRequests deletes all requests that have ended before the rate limiter's window.
+func (rl *RateLimiter) PruneRequests() {
+	rl.mu.Lock()
+	defer rl.mu.Unlock()
+
+	// delete all the requests that have
+	cutoff := time.Now().Add(-time.Duration(rl.window) * time.Millisecond)
```

Comment:
It might be nice if rl.window was already a Duration (like I mentioned in the config stuff) for lines like this.

Reply:
removed not needed conversions to Duration